Google.Cloud.Spanner.NHibernate.Tests/NHibernateMockServerTests.cs

```diff
@@ -0,0 +1,139 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Google.Cloud.Spanner.Connection.MockServer;
+using Google.Cloud.Spanner.NHibernate.Tests.Entities;
+using Google.Cloud.Spanner.V1;
+using Google.Protobuf.WellKnownTypes;
+using NHibernate;
+using Xunit;
+
+namespace Google.Cloud.Spanner.NHibernate.Tests
+{
+    /// <summary>
+    /// Tests CRUD operations using an in-mem Spanner mock server.
+    /// </summary>
+    public class NHibernateMockServerTests : IClassFixture<NHibernateMockServerFixture>
+    {
+        private readonly NHibernateMockServerFixture _fixture;
+
+        public NHibernateMockServerTests(NHibernateMockServerFixture service)
+        {
+            _fixture = service;
+            service.SpannerMock.Reset();
+        }
+
+        [Fact]
+        public async Task VersionNumberIsAutomaticallyGeneratedOnInsertAndUpdate()
+        {
+            var insertSql = "INSERT INTO SingerWithVersion (Version, FirstName, LastName, SingerId) VALUES (@p0, @p1, @p2, @p3)";
+            _fixture.SpannerMock.AddOrUpdateStatementResult(insertSql, StatementResult.CreateUpdateCount(1L));
+
+            var singer = new SingerWithVersion { SingerId = 1L, FirstName = "Pete", LastName = "Allison" };
+
+            using (var session = _fixture.SessionFactory.OpenSession())
+            {
+                using (var transaction = session.BeginTransaction())
+                {
+                    await session.SaveAsync(singer);
+                    await session.FlushAsync();
+                    await transaction.CommitAsync();
+                }
+            }
+
+            Assert.Collection(
+                _fixture.SpannerMock.Requests.OfType<ExecuteSqlRequest>(),
+                request =>
+                {
+                    Assert.Equal(insertSql, request.Sql);
+                    // Verify that the version is 1.
+                    Assert.Equal(Value.ForString("1"), request.Params.Fields["p0"]);
+                    Assert.Equal(Value.ForString("Pete"), request.Params.Fields["p1"]);
+                    Assert.Equal(Value.ForString("Allison"), request.Params.Fields["p2"]);
+                    Assert.Equal(Value.ForString("1"), request.Params.Fields["p3"]);
+                }
+            );
+
+            _fixture.SpannerMock.Reset();
+            // Update the singer and verify that the version number is included in the WHERE clause and is updated.
+            var updateSql = "UPDATE SingerWithVersion SET Version = @p0, FirstName = @p1, LastName = @p2 WHERE SingerId = @p3 AND Version = @p4";
+            var updateAlbumSql = "UPDATE AlbumWithVersion SET SingerWithVersion = null WHERE SingerWithVersion = @p0";
```

Comment:
This statement is generated by NHibernate because the list of `Albums` of the `SingerWithVersion` entity is empty, and NHibernate wants to 'guarantee' that that is the case in the database as well. Collection mappings in NHibernate can be extremely inefficient, and this is one of those examples where NHibernate generates an update statement that is not really needed. We'll include a number of optimization considerations in the documentation of the driver, and some of them will be related to avoiding unnecessary DML statements when working with collections. We should therefore also try to keep our tests in line with that, unless we are actually testing non-recommended usage of NHibernate.

Change the `Albums` collection mapping in the `SingerWithVersionMapping` to the following:

```csharp
            Bag(x => x.Albums, c =>
            {
                c.Key(k => k.NotNullable(true));
            }, r => r.OneToMany());
```

You can then remove this update statement from this test, as NHibernate will no longer generate and execute it.